src/test/java/org/company/demo/library/libraryproject/services/implementation/BookServiceImplTest.java

```diff
@@ -0,0 +1,230 @@
+package org.company.demo.library.libraryproject.services.implementation;
+
+import org.company.demo.library.libraryproject.dto.BookDTO;
+import org.company.demo.library.libraryproject.mappers.BookMapper;
+import org.company.demo.library.libraryproject.models.Book;
+import org.company.demo.library.libraryproject.models.constants.Constants;
+import org.company.demo.library.libraryproject.models.errors.BookAlreadyExistsException;
+import org.company.demo.library.libraryproject.models.errors.BookNotFoundException;
+import org.company.demo.library.libraryproject.repositories.BookRepository;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.MockitoAnnotations;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.mock.mockito.MockBean;
+import org.springframework.http.HttpStatus;
+import org.springframework.test.context.ContextConfiguration;
+import org.springframework.test.context.junit.jupiter.SpringExtension;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+import static org.company.demo.library.libraryproject.TestUtil.*;
+import static org.junit.jupiter.api.Assertions.*;
+import static org.mockito.ArgumentMatchers.anyInt;
+import static org.mockito.Mockito.*;
+
+
+@ExtendWith(SpringExtension.class)
+@ContextConfiguration(classes = {BookServiceImpl.class})
+class BookServiceImplTest {
+
+
+    @Autowired
+    BookServiceImpl bookService;
+
+    @MockBean
+    BookRepository bookRepository;
+
+    @MockBean
+    BookMapper mapper;
+
+
+    @BeforeEach
+    public void setUp() {
+        MockitoAnnotations.openMocks(this);
+    }
+
+
+    @Test
+    void addBookAlreadyExistsTest() {
+        BookDTO bookDTO = bookDTO();
+
+        when(bookRepository.existsByBookCode(anyInt())).thenReturn(true);
+        assertThrows(BookAlreadyExistsException.class, () -> bookService.addBook(bookDTO));
+        verify(bookRepository, times(1)).existsByBookCode(anyInt());
+    }
+
+    @Test
+    void addBookTest() {
```

Comment:
this test only checks if what we mocked happens instead of all logic inside the method. Date adding is missing.